src/coreclr/src/vm/dllimport.cpp

```diff
@@ -4847,7 +4811,9 @@ void NDirect::PopulateNDirectMethodDesc(NDirectMethodDesc* pNMD, PInvokeStaticSi
         if (argit.HasRetBuffArg())
         {
             MethodTable *pRetMT = msig.GetRetTypeHandleThrowing().AsMethodTable();
-            if (IsUnmanagedValueTypeReturnedByRef(pRetMT->GetNativeSize()))
+            // The System.DateTime type itself technically doesn't have a native representation,
```

Comment:
Why is this special-casing needed now?

Reply:
Before this change, the EEClass used to also store the native size so we wouldn't have to get the layout info. However, this field was only valid when `IsBlittable()` or `HasLayout()` was true. When they were both false, it happened to be `0`, which although wrong, happened to have the expected behavior for this particular case with `System.DateTime`. This PR doesn't use that field any more and instead always goes down the layout-info path, which asserts that the type has layout. Since `System.DateTIme` is auto-layout, the assert fails.

Reply:
Can the same thing happen for other auto-layout types?

Reply:
Nope. `System.DateTime` is the only auto-layout value type that is marshalable and whose native representation can be returned as a return value. The only other auto-layout value type we allow users to marshal is `System.DateTimeOffset`, which is only allowed in the WinRT case where we always return an `HRESULT`.

Reply:
If you create wrong PInvoke signature that returns auto-layout type, are we going to throw exception for it before coming here?

Reply:
It looks like we end up throwing later. So, I've added a `HasLayout()` check here so we don't crash and instead throw an exception later when actually emitting the marshalling.